### DISCUSSION TITLE
Make state names unique in animation graph state machine

### DISCUSSION THREAD

**Discussion by shrinktofit** [cocos/cocos-engine#13135] [Show and tell]
I'm planning to make the state names in animation graph state machine unique **and non-blank**. It's required since the programming side needs a unique identifier to states(for example, operation that dynamically transition to specific state).

Obviously we might suffer from compatibility issue: we need to migrate existing animation graph to deduplicate the state names. Fortuanately, this change won't introduce programming side breaking change since we didn't expose any API about relavent to state names as far, except an internal field `__DEBUG__ID__`.

Known migration behaviour:

- States with no name(blank name) assigned would be assigned a name(by our rule).

- States with reduant names would be assigned a new name so that each name in a single state machine keeps unique.

**Comment by shrinktofit** [comment]
@holycanvas  @arsen2010 @AndyYuTian   Please take a look at this. Let me know your position at this decision.

**Comment by shrinktofit** [comment]
@AndyYuTian  I impose no constraints other than the uniqueness and non-blank. But I propose to also trim the name(remove the leading and trailling white spaces), how do you feel about this.

**Reply by AndyYuTian** [reply]
I think that is a reasonable request, empty names and same naming would only cause confuse for the users

**Comment by shrinktofit** [comment]
@AndyYuTian  An important effect that is implied by this decision is that "these names becomes meaningful to programmers". This can lead to most(but NOT ALL) programmer would use only english characters as state name. So, should we take up a display name as well?

**Reply by AndyYuTian** [reply]
Although the name is unique now, but i am not sure if using the name directly as uniqueId is a good idea, should we still generate an id for each node? and user should able to use non-english characters in the name.
